examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,150 @@
+// Package timelock provides a library for scheduling, cancelling, and
+// executing time-locked operations in Gno. It ensures that
+// operations are only carried out after a specified delay and offers
+// mechanisms for managing and verifying the status of these operations.
+// This package leverages an AVL tree for efficient management of timestamps
+// and integrates role-based access control for administrative tasks.
+//
+// Example Usage:
+//
+//  import "gno.land/p/demo/timelock"
+//  import "gno.land/p/demo/accesscontrol"
+//
+//  Initialize timelock utility with an AVL tree and access control.
+//  timestamps := avl.NewTree()
+//  adminRole := accesscontrol.NewRoleData("admin", std.Address("admin-address"))
+//  timelockUtil := timelock.NewTimelockUtil(timestamps, adminRole)
+//
+//  Schedule an operation with a delay of 60 seconds and a minimum delay of 30 seconds.
+//  id := seqid.ID()
+//  timelockUtil.Schedule(id, 60, 30)
+//
+//  Check if an operation is pending.
+//  isPending := timelockUtil.IsOperationPending(id)
+//
+//  Execute the operation when it is ready.
+//  if timelockUtil.IsOperationReady(id) {
+//      timelockUtil.Execute(id)
+//  }
+//
+//  Update the minimum delay for future operations.
+//  var minDelay int64 = 30
+//  timelockUtil.UpdateDelay(&minDelay, 45)
+
+package timelock // import "gno.land/p/demo/timelock"
+
+import (
+    "strconv"
+    "std"
+    "gno.land/p/demo/avl"
+    "gno.land/p/demo/accesscontrol"
+    "gno.land/p/demo/seqid"
+    "time"
+)
+
+// Marks operations completed
+const DONE_TIMESTAMP = 1
+
+// Represents the status of a planned operation
+type OperationState int
+
+const (
+    Unset OperationState = iota
+    Waiting
+    Ready
+    Done
+)
+
+// TimelockUtil stores the necessary parameters for the timelock operations
+type TimelockUtil struct {
+    timestamps    *avl.Tree
+    accessControl *accesscontrol.RoleData
+}
+
+// New instance of TimelockUtil
+func NewTimelockUtil(timestamps *avl.Tree, accessControl *accesscontrol.RoleData) *TimelockUtil {
+    return &TimelockUtil{
+        timestamps:    timestamps,
+        accessControl: accessControl,
+    }
+}
+
+// Schedules an operation to be carried out after a minimum delay
+func (tl *TimelockUtil) Schedule(id seqid.ID, delay int64, minDelay int64) {
```

Comment:
Can you explain the purpose of `minDelay`? If this value is greater than `delay`, then it seems like this method shouldn't be called in the first place.  Removing `minDelay` also makes it easier for the reader to ascertain what this method does just by looking at the signature.

Reply:
minDelay is used to ensure a minimum delay before an operation is executed, guaranteeing safety and enabling users to react to critical changes. delay is used to customize specific delays while respecting minDelay, offering planning flexibility while preventing errors and abuse. The structure of this method is inspired by openzeppelin [TimeLockController.](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/TimelockController.sol)

Reply:
Okay, I see. In that solidity code the `minDelay` is a field on the `TimeLockController`; it isn't passed it when calling each method. Can the same be done here by making `minDelay` a field on `TimelockUtil` rather than passing it in with each method call?

Reply:
You're right, I made this modification to make things more coherent.

- https://github.com/gnolang/gno/pull/2307/commits/59432e040ee95ef8d325b6268cf4ffc1981095be